src/esplog/mini_decompressor.rs

```diff
@@ -0,0 +1,691 @@
+    }
+
+    pub fn sdiv(&self, x: Fix) -> Quat {
+        Quat {
+            w: self.w / x,
+            x: self.x / x,
+            y: self.y / x,
+            z: self.z / x,
+        }
+    }
+}
+
+impl Add for &Quat {
+    type Output = Quat;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        Quat {
+            w: self.w + rhs.w,
+            x: self.x + rhs.x,
+            y: self.y + rhs.y,
+            z: self.z + rhs.z,
+        }
+    }
+}
+
+impl Mul for &Quat {
+    type Output = Quat;
+
+    fn mul(self, rhs: Self) -> Self::Output {
+        Quat {
+            w: self.w * rhs.w - self.x * rhs.x - self.y * rhs.y - self.z * rhs.z,
+            x: self.w * rhs.x + self.x * rhs.w + self.y * rhs.z - self.z * rhs.y,
+            y: self.w * rhs.y - self.x * rhs.z + self.y * rhs.w + self.z * rhs.x,
+            z: self.w * rhs.z + self.x * rhs.y - self.y * rhs.x + self.z * rhs.w,
+        }
+    }
+}
+
+impl Add for Quat {
+    type Output = Quat;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        &self + &rhs
+    }
+}
+
+impl Mul for Quat {
+    type Output = Quat;
+
+    fn mul(self, rhs: Self) -> Self::Output {
+        &self * &rhs
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct Vec3<const N: usize> {
+    pub x: Fix32<N>,
+    pub y: Fix32<N>,
+    pub z: Fix32<N>,
+}
+
+impl<const N: usize> Vec3<N> {
+    pub fn new(x: Fix32<N>, y: Fix32<N>, z: Fix32<N>) -> Vec3<N> {
+        Vec3 { x, y, z }
+    }
+
+    pub fn smul(&self, k: Fix32<N>) -> Vec3<N> {
+        Vec3 {
+            x: self.x * k,
+            y: self.y * k,
+            z: self.z * k,
+        }
+    }
+
+    pub fn sdiv(&self, k: Fix32<N>) -> Vec3<N> {
+        Vec3 {
+            x: self.x / k,
+            y: self.y / k,
+            z: self.z / k,
+        }
+    }
+
+    pub fn norm(&self) -> Fix32<N> {
+        (self.x * self.x + self.y * self.y + self.z * self.z).sqrt()
+    }
+
+    pub fn normalized(&self) -> Vec3<N> {
+        self.sdiv(self.norm())
+    }
+}
+
+impl<const N: usize> Add for Vec3<N> {
+    type Output = Vec3<N>;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        Vec3 {
+            x: self.x + rhs.x,
+            y: self.y + rhs.y,
+            z: self.z + rhs.z,
+        }
+    }
+}
+
+impl<const N: usize> Sub for Vec3<N> {
+    type Output = Vec3<N>;
+
+    fn sub(self, rhs: Self) -> Self::Output {
+        Vec3 {
+            x: self.x - rhs.x,
+            y: self.y - rhs.y,
+            z: self.z - rhs.z,
+        }
+    }
+}
+
+impl<const N: usize> Neg for Vec3<N> {
+    type Output = Vec3<N>;
+
+    fn neg(self) -> Self::Output {
+        Vec3 {
+            x: -self.x,
+            y: -self.y,
+            z: -self.z,
+        }
+    }
+}
+
+impl<const N: usize> Default for Vec3<N> {
+    fn default() -> Self {
+        Self {
+            x: Fix32::<N>::from_i32(0),
+            y: Fix32::<N>::from_i32(0),
+            z: Fix32::<N>::from_i32(0),
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
```

Comment:
by adding `Default` to `derive` you can make `new` simply `pub fn new() -> Self { Self::default() }`